File: android/core/src/main/kotlin/com/uwbcompass/core/BleDistanceModel.kt
package com.uwbcompass.core

import kotlin.math.pow

/**
 * Log-distance path-loss RSSI → distance model, shared by the BLE provider and its tests.
 * Pure so it is unit-tested without hardware.
 *
 *   distance = 10 ^ ((txPowerAt1m - rssi) / (10 * pathLossExponent))
 *
 * [txPowerDbm] is the measured RSSI at 1 m (VERIFY-ON-DEVICE per phone); [pathLossExponent]
 * is ~2.0 in free space, higher indoors.
 */
object BleDistanceModel {
    fun rssiToDistanceMeters(rssi: Int, txPowerDbm: Int = -59, pathLossExponent: Double = 2.0): Double =
        10.0.pow((txPowerDbm - rssi) / (10.0 * pathLossExponent))

    fun qualityFrom(distanceMeters: Double): SignalQuality = when {
        distanceMeters <= 3.0 -> SignalQuality.HIGH
        distanceMeters <= 10.0 -> SignalQuality.MEDIUM
        else -> SignalQuality.LOW
    }
}

File: ios/Sources/UWBCompassCore/BleDistanceModel.swift
import Foundation

/// Log-distance path-loss RSSI → distance model, mirror of the Android one. Pure & tested.
///   distance = 10 ^ ((txPowerAt1m - rssi) / (10 * pathLossExponent))
public enum BleDistanceModel {
    public static func rssiToDistanceMeters(rssi: Int, txPowerDbm: Int = -59, pathLossExponent: Double = 2.0) -> Double {
        pow(10.0, Double(txPowerDbm - rssi) / (10.0 * pathLossExponent))
    }

    public static func quality(forDistance meters: Double) -> SignalQuality {
        if meters <= 3 { return .high }
        if meters <= 10 { return .medium }
        return .low
    }
}
